azure/services/aso/tags.go

```diff
@@ -47,6 +50,14 @@ func reconcileTags[T genruntime.MetaObject](t TagsGetterSetter[T], existing T, r
 		}
 
 		existingTags = t.GetActualTags(existing)
+		if !reflect.DeepEqual(t.GetDesiredTags(existing), existingTags) &&
```

Comment:
I had noticed a race condition while running the tags e2e test where if the tags have not converged, then CAPZ could lose track of some tags that have been deleted from `spec.additionalTags` and treat them like tags that existed outside of `spec.additionalTags` and keep them around without ever deleting them. Waiting for the tags to converge on the ASO resource fixes that issue AFAICT.

Overall, this makes tags reconciliation work more similarly to the more synchronous method used before when calling the tags APIs directly.

Reply:
Merged in #4149